_posts/2023-05-17-Scaling_part_2.md

```diff
@@ -0,0 +1,642 @@
+---
+title: "Scaling to 1000 clusters - Part 2"
+date: 2023-05-17
+draft: false
+categories: ["metal3", "cluster API", "provider", "edge"]
+author: Lennart Jern
+---
+
+In [part 1]({% link _posts/2023-05-05-Scaling_part_1.md %}), we introduced the Bare Metal Operator test mode and saw how it can be used to play with BareMetalHosts without Ironic and without any actual hosts.
+Now we will take a look at the other end of the stack and how we can fake the workload cluster API's.
+
+## Test setup
+
+The end goal is to have one management cluster where the Cluster API and Metal3 controllers run.
+In this cluster we would generate BareMetalHosts and create Clusters, Metal3Clusters, etc to benchmark the controllers.
+To give them a realistic test, we also need to fake the workload cluster API's.
+These will run separately to avoid interfering with the test (e.g. by using up all the resources in the management cluster).
+Here is a diagram that describes the setup:
+
+![diagram of test setup](/assets/2023-05-17-Scaling_part_2/scaling-fake-clusters.drawio.png)
+
+How are we going to fake the workload cluster API's then?
+The most obvious solution is to just run the real deal, i.e. the `kube-apiserver`.
+This is what would be run in a real workload cluster, together with the other components that makes up the Kubernetes control plane.
+
+## Running the Kubernetes API server
+
+There are many misconceptions, maybe even superstition, about the Kubernetes control plane.
+The fact is that it is in no way special.
+It consists of a few programs that can be run in any way you want: in a container, as a systemd unit or directly executed at the command line.
+They can run on a Node or outside of the cluster.
+You can even run multiple instances on the same host as long as you avoid port collisions.
+
+For our purposes we basically want to run as little as possible of the control plane components.
+We just need the API to be available and possible for us to populate with data that the controllers expect to be there.
+In other words, we need the API server and etcd.
+The scheduler is not necessary since we won't run any actual workload (we are just pretending the Nodes are there anyway) and the controller manager would just get in the way when we want to fake resources.
+It would, for example, try to update the status of the (fake) Nodes that we want to create.
+
+The API server will need an etcd instance to connect to.
+It will also need some TLS configuration, both for connecting to etcd and for handling service accounts.
+One simple way to generate the needed certificates is to use kubeadm.
+But before we get there we need to think about how the configuration should look like.
+
+For simplicity, we will simply run the API server and etcd in a kind cluster for now.
+It would then be easy to run them in some other Kubernetes cluster later if needed.
+Let's create it right away:
+
+```bash
+kind create cluster
```

Comment:
Would be nice to set explicitly the k8s version we are using in this blog to avoid unexpected results or compatibility issues

Reply:
Missed this the first time. Added now.
There are of course many more versions that we could specify but I would prefer to not go so deep with details. There is no way to make this completely reproducible without freezing all packages on the host and so on :slightly_frowning_face:

Reply:
Agree but my UX here was not good since first issue I hit when trying this was running kind with newer version and one of our file was specifying 1.25.x which caused some incompatibility

Reply:
Ow that's bad. Let me know if I need to put more versions somewhere :)